server/rest/src/main/java/org/infinispan/rest/server/operations/CacheOperations.java

```diff
@@ -0,0 +1,303 @@
+            if (meta instanceof MimeMetadata) {
+               String etag = calcETAG(ice.getValue(), ((MimeMetadata) meta).contentType());
+               Optional<String> clientEtag = request.getEtag();
+               if(clientEtag.map(t -> t.equals(etag)).orElse(true)) {
+                  if (useAsync.isPresent() && useAsync.get()) {
+                     restCacheManager.getCache(cacheName).removeAsync(key);
+                  } else {
+                     restCacheManager.getCache(cacheName).remove(key);
+                  }
+               } else {
+                  //ETags don't match, so preconditions failed
+                  response.status(HttpResponseStatus.PRECONDITION_FAILED);
+               }
+            }
+         }
+         return response;
+      } catch (CacheException cacheException) {
+         throw new NoCacheFoundException(cacheException.getLocalizedMessage());
+      }
+   }
+
+   public InfinispanResponse deleteEntireCache(InfinispanRequest request) throws RestResponseException {
+      try {
+         String cacheName = request.getCacheName();
+         Optional<Boolean> useAsync = request.getUseAsync();
+
+         InfinispanResponse response = InfinispanResponse.inReplyTo(request);
+         response.status(HttpResponseStatus.OK);
+
+         if (useAsync.isPresent() && useAsync.get()) {
+            restCacheManager.getCache(cacheName).clearAsync();
+         } else {
+            restCacheManager.getCache(cacheName).clear();
+         }
+
+         return response;
+      } catch (CacheException cacheException) {
+         throw new NoCacheFoundException(cacheException.getLocalizedMessage());
+      }
+   }
+
+   private boolean supportsExtendedHeaders(String extended) {
+      switch (restServerConfiguration.extendedHeaders()) {
+         case NEVER:
+            return false;
+         case ON_DEMAND:
+            return extended != null;
+         default:
+            return false;
+      }
+   }
+
+   private CacheControl calcCacheControl(Date expires) {
+      if (expires == null) {
+         return null;
+      }
+      int maxAgeSeconds = calcFreshness(expires);
+      if (maxAgeSeconds > 0)
+         return CacheControl.maxAge(maxAgeSeconds);
+      else
+         return CacheControl.noCache();
+   }
+
+
+   private boolean entryFreshEnough(Date entryExpires, OptionalInt minFresh) {
+      return !minFresh.isPresent() || minFresh.getAsInt() < calcFreshness(entryExpires);
+   }
+
+   private int calcFreshness(Date expires) {
+      if (expires == null) {
+         return Integer.MAX_VALUE;
+      } else {
+         return ((int) (expires.getTime() - new Date().getTime()) / 1000);
+      }
+   }
+
+   private OptionalInt minFresh(String cacheControl) {
+      Optional<String> minFreshDirective = Arrays.stream(cacheControl.split(",")).filter(s -> s.contains("min-fresh")).findFirst();
+      return minFreshDirective.map(s -> {
+         String[] equals = s.split("=");
```

Comment:
Seems you could do all this map stuff in the above stream. It is done lazily so it would only be performed on the first match anyways.